File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {

    int m;
    int n;

    public List<Integer> spiralOrder(int[][] matrix) {
        int[] dx = new int[] {0, 1, 0, -1};
        int[] dy = new int[] {1, 0, -1, 0};
        this.m = matrix.length;
        this.n = matrix[0].length;
        
        boolean[][] visited = new boolean[this.m][this.n];
        List<Integer> result= new ArrayList<>();
        
        int end = 1;
        int x = 0, y = 0;
        int idx = 0;
        result.add(matrix[0][0]);
        
        while (true) {
            if (end == this.m*this.n) break;

            visited[x][y] = true;
            
            int new_x = x + dx[idx%4];
            int new_y = y + dy[idx%4];
            
            
            if (go_outside(new_x, new_y)) {
                idx++;
                continue;
            }
            
            if (visited[new_x][new_y]) {
                idx++;
                continue;
            }
            result.add(matrix[new_x][new_y]);
            x = new_x;     
            y = new_y;
            end++;
        }
        
        return result;
        }
    
    
    public boolean go_outside(int i, int j) {
        if (i < 0 || i >= this.m || j < 0 || j >= this.n) {
            return true;
        }
        return false;
    }

}

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        
        n = len(matrix)
        m = len(matrix[0])
        
        def go_outside(x, y):
            if x < 0 or x >= n or y < 0 or y >= m:
                return True
            return False
        
        visited = [[False] * m for _ in range(n)]
        
        end = 1
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        
        x = 0
        y = 0
        
        ans = [matrix[0][0]]
        visited[0][0] = True
        
        idx = 0
        while (True):
            if (end == n*m):
                break
            
            dx = directions[idx%4][0]
            dy = directions[idx%4][1]
            
            if (go_outside(x+dx, y+dy)):
                idx += 1
                continue
            if (visited[x+dx][y+dy]):
                idx += 1
                continue
            x = x + dx
            y = y + dy
            ans.append(matrix[x][y])
            visited[x][y] = True
            end += 1
            
        return ans
            

        